File: frontend/src/lib/dataTrust.ts
// Types plus a dev-only fixture for the Data-Trust strip: the thin always-visible bar (in the Shell)
// that tells you, on every page, how much to trust what you are looking at. Freshness of the account
// snapshot, the price source and how many positions actually priced, whether returns are price-only,
// and the two posture facts that change how much the numbers mean (auth enforced, live debates on).
//
// Mirrors the PROPOSED `GET /api/data-trust` contract in docs/contracts/data-trust-endpoint.md. The
// backend already has every field: the account snapshot (generated_at, source, stale_prices, per-
// position `priced`) and /api/health (debate_ready, auth_enforced). This route just bundles them so
// the strip is one cheap fetch that does NOT pull the whole portfolio on pages that don't need it.
//
// Not in lib/types.ts on purpose (same reason as perf/calibration/reconciliation/position): that
// file mirrors shapes a real backend returns today, and this endpoint does not exist yet.

import type { ReturnsBasis } from "@/lib/perf";

export interface DataTrustResponse {
  // ── Freshness (from the account snapshot) ──
  snapshot_generated_at: string | null;
  /** Snapshot older than the freshness window; the operator should refresh before acting. */
  snapshot_stale: boolean;
  // ── Pricing (from the live FMP marks) ──
  price_source: string;
  /** At least one held symbol could not be priced live (account.stale_prices). */
  prices_degraded: boolean;
  positions_total: number;
  positions_priced: number;
  /** price_only is a standing honesty caveat: dividends/total-return are not in these numbers. */
  returns_basis: ReturnsBasis;
  // ── Posture (from /api/health): changes how much the numbers above can be trusted ──
  /** Live debates are on (cost API tokens) vs the debate pages being history-only. */
  debate_live: boolean;
  /** Protected routes require an operator session. False is the legitimate pre-auth posture, but it
   *  is also what a mislaid backend/.env looks like, so it is shown, never inferred. */
  auth_enforced: boolean;
}

// A single boolean the strip can trust: true when ANY page in the app is showing a fixture instead
// of real data. Read straight from the build-time NEXT_PUBLIC_*_MOCK flags (Next inlines them), so
// the strip can say "some views are mock" without asking the backend. Keep this list in sync as new
// mock-gated pages are added.
export const ANY_MOCK =
  process.env.NEXT_PUBLIC_PERF_MOCK === "1" ||
  process.env.NEXT_PUBLIC_CALIB_MOCK === "1" ||
  process.env.NEXT_PUBLIC_RECON_MOCK === "1" ||
  process.env.NEXT_PUBLIC_POSITION_MOCK === "1" ||
  process.env.NEXT_PUBLIC_MARKET_MOCK === "1" ||
  process.env.NEXT_PUBLIC_TESTLAB_MOCK === "1" ||
  process.env.NEXT_PUBLIC_TRUST_MOCK === "1";

// ── Dev fixture ────────────────────────────────────────────────────────────────────────────────
// Strict gate (NEXT_PUBLIC_TRUST_MOCK=1); the default hits the real endpoint. The numbers match the
// reconciliation/position fixtures: a stale (2026-07-27) snapshot, 8 held names with 2 that couldn't
// be priced, price-only returns, live debates off, and (for the demo) auth enforced.
export const TRUST_MOCK = process.env.NEXT_PUBLIC_TRUST_MOCK === "1";

export const MOCK_DATA_TRUST: DataTrustResponse = {
  snapshot_generated_at: "2026-07-27T15:30:00Z",
  snapshot_stale: true,
  price_source: "FMP",
  prices_degraded: true,
  positions_total: 8,
  positions_priced: 6,
  returns_basis: "price_only",
  debate_live: false,
  auth_enforced: true,
};

File: backend/app/routers/data_trust.py
"""GET /api/data-trust — can you believe the numbers on this page?

Feeds the always-visible strip the Shell renders above every page
(``frontend/src/components/data-trust.tsx``; contract in ``docs/contracts/data-trust-endpoint.md``).

WHY THIS ROUTE EXISTS
    For three weeks the Portfolio page showed holdings dated 27 July as though they were current.
    Prices were live, which made it harder to notice rather than easier — the numbers moved. Nothing
    on the screen said the positions underneath them were frozen, and the only reason anyone found
    out was that the owner asked.

    This strip is the thing that says so. It is deliberately small and deliberately pessimistic: it
    reports what it can prove and refuses to render a green state it has not earned.

CHEAP BY CONSTRUCTION
    It renders on Scan, Pipeline and Debate — pages that never show a portfolio. Proxying
    /api/account would make each of them pay for a full portfolio payload plus a broker round-trip
    for a status bar. So the snapshot is read once here (through the same short-TTL broker cache the
    account view uses) and priced against the same marks cache. No new upstream calls.

NOT A HEALTH CHECK
    /api/health answers "is the service up" for a container probe. This answers "is what you are
    looking at true", which is a different question with different consumers. They share two posture
    flags and nothing else.
"""

from __future__ import annotations

import logging
from datetime import datetime, timezone
from typing import Any

from fastapi import APIRouter

from app.config import get_settings
from app.services.broker import get_snapshot
from app.services.marks import MARKS_PROVIDER, get_marks_detailed, resolve_ttl_seconds
from app.services.snapshot import SnapshotError

logger = logging.getLogger("agentic.api.data_trust")

router = APIRouter(prefix="/api", tags=["data-trust"])

# Until dividends and corporate actions flow into the return series, every performance figure in
# this app is price-only. Surfaced as a standing caveat rather than a footnote nobody reads.
RETURNS_BASIS = "price_only"


def _parse_iso(value: str | None) -> datetime | None:
    """Parse the snapshot's ISO-8601 stamp, tolerating the trailing Z."""
    if not value:
        return None
    try:
        return datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        logger.warning("snapshot generated_at is not ISO-8601: %r", value)
        return None


def _budget_status() -> dict[str, Any]:
    """How much of the daily provider allowance this process has spent.

    Process-local and approximate by construction (src/fmp.py::CallBudget) — the key's responses
    carry no rate-limit headers. Approximate and visible beats exact and absent: the previous
    budget was sized for a backfill, the dashboard quietly outspent it every ten minutes, and
    nothing anywhere reported that until positions started disappearing.
    """
    try:
        from src.fmp import get_shared_client

        budget = get_shared_client().budget
        limit = int(budget.limit)
        spent = int(budget.spent)
    except Exception:  # noqa: BLE001 — a reporting failure must not fail the page
        return {"spent": None, "limit": None, "exhausted": None}
    return {
        "spent": spent,
        # 0 means no daily cap is configured, which is not the same as a cap of zero.
        "limit": limit or None,
        "exhausted": bool(limit and spent >= limit),
    }


@router.get("/data-trust")
def data_trust() -> dict[str, Any]:
    """Freshness, pricing coverage, and posture — the whole strip in one small payload."""
    settings = get_settings()
    # Imported here rather than at module scope so the strip reports auth posture through the same
    # function /api/health uses, without the two routers importing each other.
    from app.services.auth import auth_enforcement_configured

    posture: dict[str, Any] = {
        "returns_basis": RETURNS_BASIS,
        "debate_live": settings.anthropic_api_key is not None,
        "auth_enforced": auth_enforcement_configured(),
    }

    try:
        snapshot = get_snapshot(settings.snapshot_path)
    except SnapshotError as exc:
        # 200 with an honest empty state, NOT an error. "We have no fresh data" is a fact the strip
        # must be able to render; it is a different thing from the endpoint being unreachable, and
        # collapsing the two would leave the operator unable to tell "broker down" from "site down".
        logger.warning("data-trust: no account data available: %s", exc)
        return {
            "snapshot_generated_at": None,
            "snapshot_stale": True,
            "source": None,
            "price_source": MARKS_PROVIDER,
            "prices_degraded": True,
            "positions_total": 0,
            "positions_priced": 0,
            **posture,
        }

    generated = _parse_iso(snapshot.generated_at)
    if generated is None:
        # An unparseable stamp cannot be shown to be fresh, so it is stale. Freshness must be
        # PROVEN, never assumed — assuming it is how the July snapshot went unremarked.
        stale = True
    else:
        age = (datetime.now(timezone.utc) - generated).total_seconds()
        stale = age > settings.snapshot_max_age_seconds

    symbols = snapshot.symbols
    ttl = resolve_ttl_seconds(settings.marks_ttl_seconds)
    detailed = get_marks_detailed(symbols, ttl) if symbols else {}
    priced = sum(1 for s in symbols if detailed.get(s) and detailed[s].price is not None)
    # Named, not just counted. "9/10 priced" leaves the operator hunting for the missing one, and a
    # stale-but-served price is a different state from an absent one — both were invisible during
    # the 2026-08-19 outage, when every position silently went unpriced at once.
    unpriced = sorted(s for s in symbols if not (detailed.get(s) and detailed[s].price is not None))
    stale_served = sorted(s for s in symbols if detailed.get(s) and detailed[s].stale)

    return {
        "snapshot_generated_at": snapshot.generated_at,
        "snapshot_stale": stale,
        # Which account this is. The dashboard shows one at a time and the source changed under
        # everyone once already; leaving it to be inferred from context is how that goes unnoticed.
        "source": snapshot.source,
        "price_source": MARKS_PROVIDER,
        "prices_degraded": priced < len(symbols),
        "positions_total": len(symbols),
        "positions_priced": priced,
        "unpriced_symbols": unpriced,
        # Priced, but from a cache the provider would not refresh. Real numbers, older than the
        # refresh interval, and the page must be able to say so rather than showing them as live.
        "stale_priced_symbols": stale_served,
        "price_refresh_seconds": ttl,
        # The quota that took the dashboard down. Surfaced so exhaustion is visible while there is
        # still headroom, instead of arriving as a blank table.
        "provider_budget": _budget_status(),
        **posture,
    }
